PreProcessing/updateExport3DPositionsVelocities.m

```diff
@@ -0,0 +1,265 @@
+            || isempty(S.bounds.distanceConstraints(i).direction)
+        S.bounds.distanceConstraints(i).direction = 'xyz';
+    end
+
+    % constraint needs at least 1 bound
+    hasLowerBound = isfield(S.bounds.distanceConstraints(i),'lower_bound')...
+            && ~isempty(S.bounds.distanceConstraints(i).lower_bound);
+    hasUpperBound = isfield(S.bounds.distanceConstraints(i),'upper_bound')...
+            && ~isempty(S.bounds.distanceConstraints(i).upper_bound);
+    if ~hasLowerBound && ~hasUpperBound
+        error(['The distance constraint between "',...
+            S.bounds.distanceConstraints(i).point1, ...
+            '" and "' S.bounds.distanceConstraints(i).point2,...
+            '" ("S.bounds.distanceConstraints(',...
+            num2str(i) ')") has no value for upper or lower bound.'])
+    end
+
+
+end
+
+%% Remove duplicate points
+segments = S.bounds.points;
+isUnique = true(size(segments));
+
+for ii = 1:length(segments)-1
+    for jj = ii+1:length(segments)
+        if isequal(segments(ii),segments(jj))
+            isUnique(jj) = false;
+            break;
+        end
+    end
+end
+
+segments(~isUnique) = [];
+
+S.bounds.points = segments;
+
+%% Add points for constraints to export3DPositions
+
+S.OpenSimADOptions.export3DPositions =...
+    [S.OpenSimADOptions.export3DPositions(:), S.bounds.points(:)];
+
+
+%% Test that the bodies exist in the osim model
+
+model = Model(osim_path);
+bodyset = model.getBodySet();
+
+for j=["export3DPositions","export3DVelocities"]
+    
+    segments = S.OpenSimADOptions.(j);
+    
+    if isempty(segments)
+        continue
+    end
+
+    validNames = [];
+    
+    for i=1:length(segments)
+        try
+            body = bodyset.get(segments(i).body);
+            validNames(end+1) = i;
+        catch
+            if ~strcmpi(segments(i).body,'ground') % no warning for removing ground
+                warning(['   Cannot find body name "' segments(i).body,...
+                    '" in osim model. Removing from S.OpenSimADOptions.',char(j),'.'])
+            end
+        end
+    end
+    
+    segments = segments(validNames);
+
+    % Remove duplicate entries
+    isUnique = true(size(segments));
+
+    for ii = 1:length(segments)-1
+        for jj = ii+1:length(segments)
+            if isequal(segments(ii),segments(jj))
+                isUnique(jj) = false;
+                break;
+            end
+        end
+    end
+    
+    segments(~isUnique) = [];
+
+    S.OpenSimADOptions.(j) = segments;
+end
+
+%% Remove constraints that rely on undefined points
+
+validConstraints = true(length(S.bounds.distanceConstraints),1);
+pointNames = [];
+for i=1:length(S.OpenSimADOptions.export3DPositions)
+    pointNames{end+1} = S.OpenSimADOptions.export3DPositions(i).name;
+end
+for i=1:length(S.bounds.points)
+    if strcmpi(S.bounds.points(i).body,'ground')
+        % add points in ground to constraints we want to keep
```

Comment:
Does that mean that only the points defined in ground will be used? Is the else missing here?

Reply:
I added some comments in the file to clarify what's going on. In short:

- 'ground' is not a body in OpenSim
- we do not want to ask OpenSimAD what the position in ground is for a point attached to the ground.